### PR TITLE
Ensure that we delist the trigger session on exception in the BeforeSave cycle

### LINKED ISSUE #168: Triggers not firing intermittently
**Version**
3.2.1
Running in net7.0 aspnet application

**Scenario**

Some code omitted for brevity. 

```
    public class SubmitFleetBookingForReviewHandler
        : IRequestHandler<SubmitFleetBookingForReviewInput, SubmitFleetBookingForReviewOutput>
    {
        private readonly ILogger<SubmitFleetBookingForReviewHandler> _logger;
        private readonly IRequestAuthVariables _requestAuthVars;
        private readonly ISupplierAuthorizer _supplierAuthorizer;
        private readonly IClock _clock;
        private readonly AutoGuruDbContext _db;
        private readonly IBookingStatusTransitionValidator _bookingStatusTransitioner;
        private readonly IRequestHandler<AcceptFleetSignOnAgreementRequest, Unit> _acceptFleetSignOnAgreementHandler;
        private readonly IRequestHandler<AssessBookingBusinessRulesRequest, BookingBusinessRuleBookingAssessment> _assessBookingBusinessRulesHandler;

        public SubmitFleetBookingForReviewHandler(
            AutoGuruDbContext db,
        )
        {
            _db = db;
        }

        public async Task<SubmitFleetBookingForReviewOutput> Handle(
            SubmitFleetBookingForReviewInput request,
            CancellationToken cancellationToken)
        {
            <code here that changes the booking>

            try
            {
                _logger.LogDebug("Before 1st save. Booking Id = {bookingId} Booking Status Id = {bookingStatusId} DbContextId = {dbContextId} Lease = {lease}",
                    bookingId, booking.BookingStatusID, _db.ContextId.InstanceId, _db.ContextId.Lease);
                _logger.LogDebug("Changes before 1st save. {changes}", _db.ChangeTracker.DebugView.LongView);

                await _db.SaveChangesAsync(CancellationToken.None);

                _logger.LogDebug("After 1st save. Booking Id = {bookingId} Booking Status Id = {bookingStatusId} DbContextId = {dbContextId} Lease = {lease}",
                    bookingId, booking.BookingStatusID, _db.ContextId.InstanceId, _db.ContextId.Lease);
                _logger.LogDebug("Changes after 1st save. {changes}", _db.ChangeTracker.DebugView.LongView);

                _logger.LogInformation("Fleet booking ({bookingId}) submitted for review,", bookingId);
            }
            catch (DbUpdateConcurrencyException)
            {
                return new SubmitFleetBookingForReviewOutput(bookingId, new ConcurrentUpdateError());
            }

            <code here that changes the booking again>

            _logger.LogDebug("Before 2nd save. Booking Id = {bookingId} Booking Status Id = {bookingStatusId} DbContextId = {dbContextId} Lease = {lease}",
                bookingId, booking.BookingStatusID, _db.ContextId.InstanceId, _db.ContextId.Lease);
            _logger.LogDebug("Changes before 2nd save. {changes}", _db.ChangeTracker.DebugView.LongView);

            await _db.SaveChangesAsync(CancellationToken.None);

            _logger.LogDebug("After 2nd save. Booking Id = {bookingId} Booking Status Id = {bookingStatusId} DbContextId = {dbContextId} Lease = {lease}",
                bookingId, booking.BookingStatusID, _db.ContextId.InstanceId, _db.ContextId.Lease);
            _logger.LogDebug("Changes after 2nd save. {changes}", _db.ChangeTracker.DebugView.LongView);
        }
    }
```

We have a number of triggers attached to the `Booking`, some `IBeforeSaveTrigger` and some `IAfterSaveTrigger`.

Intermittently the triggers on `Booking` don't fire during the 2nd `SaveChangesAsync()` in the code above.

**Logs**

[FailedLogs.txt](https://github.com/koenbeuk/EntityFrameworkCore.Triggered/files/10173611/FailedLogs.txt)

[SuccessLogs.txt](https://github.com/koenbeuk/EntityFrameworkCore.Triggered/files/10173629/SuccessLogs.txt)

**Differences**
When successful, 33 triggers are discovered.  When it fails, the first 22 triggers match, but 11 are missing.

```
Invoking trigger: "AutoGuru.DataAccess.Triggers.Booking_BeforeSave_SetSupplierPaidAt" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Invoking trigger: "AutoGuru.DataAccess.Triggers.Booking_BeforeSave_SetAutoGuruPaidAt" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Invoking trigger: "AutoGuru.DataAccess.Triggers.Booking_BeforeSave_SetRatesAndDiscountsForFleetBookings" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Invoking trigger: "AutoGuru.DataAccess.Triggers.Booking_BeforeSave_SetPayerAndPaymentTerms" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Booking_BeforeSave_SetMerchantServiceFeeForFleetBooking. Booking Id = 6005225 % = 0.0350 $ = 14.37
Booking Id = 6005225 Fleet Company Id = 1 Supplier Id = 830 Booking Status Id = 33 DbContextId = fba0254d-8bb0-4555-adff-a648aa665397 Lease 26
Invoking trigger: "AutoGuru.DataAccess.Triggers.Booking_BeforeSave_SetMerchantServiceFeeForFleetBooking" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Invoking trigger: "AutoGuru.DataAccess.Triggers.Booking_BeforeSave_SetLastReviewedPropertiesOfBookingTicketItems" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Invoking trigger: "AutoGuru.DataAccess.Triggers.Booking_BeforeSave_SetDataFee" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Invoking trigger: "AutoGuru.DataAccess.Triggers.Booking_BeforeSave_ResetClaimOnAction" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Invoking trigger: "AutoGuru.Shared.Bookings.Booking_ConsumeSupplierTimeSlot" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Booking]"
Invoking trigger: "AutoGuru.DataAccess.Triggers.CreationTimestampingTrigger" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Abstractions.ICreationTimestamped]"
Invoking trigger: "AutoGuru.DataAccess.Triggers.UpdateTimestampingTrigger" as "EntityFrameworkCore.Triggered.IBeforeSaveTrigger`1[AutoGuru.Entities.Abstractions.IUpdateTimestamped]"
```

The logs show the DbContextId of the DbContext injected into the code above and the DbContextId used in the trigger discovery process (`TriggerContextTracker`). The same context is used (as expected) but the changes don't seem to be detected.

We're using the default cascade behavior and limit.

In terms of testing, it is sometimes very intermittent in that I could use JMeter to submit 1000 bookings and all are successful. In other tests, multiple could fail within the first 100 requests.

I will be doing further investigations and will update as I make discoveries.

### PR DESCRIPTION
**Pull request by koenbeuk**
Closes #168

### CODE REVIEW

File: src/EntityFrameworkCore.Triggered/Internal/TriggerSessionSaveChangesInterceptor.cs [hunk 1]
```diff
@@ -80,14 +80,19 @@ public InterceptionResult<int> SavingChanges(DbContextEventData eventData, Inter
...
 
                 _triggerSession.RaiseBeforeSaveStartingTriggers();
                 _triggerSession.RaiseBeforeSaveTriggers();
                 _triggerSession.CaptureDiscoveredChanges();
                 _triggerSession.RaiseBeforeSaveCompletedTriggers();
             }
+            catch
+            {
+                // We're aborting the SaveChanges call, delist the trigger session now
+                DelistTriggerSession(eventData);
```
Comment (anthony-keller) [inline src/EntityFrameworkCore.Triggered/Internal/TriggerSessionSaveChangesInterceptor.cs:93]: Would this bed to be applied to the async version of this method as well?

Comment (koenbeuk) [reply src/EntityFrameworkCore.Triggered/Internal/TriggerSessionSaveChangesInterceptor.cs:93]: Nope, just added the async fix as well. thanks!

**Comment by anthony-keller** [comment]
I've done some testing after applying these changes manually to the code I've been using and have encountered a different issue.

We have an integration test:

```
                var booking = _seeder.SuppSrcedBnplBooking;

                var effectiveAt = new DateTimeOffset(2021, 11, 11, 3, 0, 0, TimeSpan.Zero);

                var ex1 = await Should.ThrowAsync<NotImplementedException>(() =>
                    PaymentGenerator.InsertPaymentAsync(db, PaymentMethodEnum.Afterpay.ToString(), 100m, null, effectiveAt));
                ex1.Message.ShouldBe("Couldn't find any applicable payment method for given args. " +
                    $"EffectiveAt: {effectiveAt}, PaymentMethodId: {PaymentMethodEnum.Afterpay}, BookingTypeId: , SupplierId: , SupplierCommissionPercentage: ");

                var ex2 = await Should.ThrowAsync<NotImplementedException>(() =>
                    PaymentGenerator.InsertPaymentAsync(db, PaymentMethodEnum.Afterpay.ToString(), 100m, booking.BookingID, effectiveAt));
                ex2.Message.ShouldBe("Couldn't find any applicable payment method for given args. " +
                    $"EffectiveAt: {effectiveAt}, PaymentMethodId: {PaymentMethodEnum.Afterpay}, BookingTypeId: {booking.BookingTypeID}, SupplierId: {booking.SupplierID}, SupplierCommissionPercentage: 0.00000");
```

The save changes fires a before save trigger:

```
        public async Task BeforeSave(ITriggerContext<Payment> context, CancellationToken cancellationToken)
        {
            if (context.ChangeType != ChangeType.Added)
            {
                return;
            }

            var payment = context.Entity;

            var fees = await _paymentMerchantFeesService.CalculateFeesAsync(
                payment.PaymentMethodId,
                payment.AmountIncGst,
                payment.BookingID,
                payment.CreatedAt,
                cancellationToken);

            payment.MerchantFeeExGst = fees.MerchantFeeExGst;
        }
```

Which eventually thows a NotImplementedException:

```
                throw new NotImplementedException(
                    "Couldn't find any applicable payment method for given args. " +
                    $"EffectiveAt: {effectiveAt}, " +
                    $"PaymentMethodId: {paymentMethodId}, " +
                    $"BookingTypeId: {bookingTypeId}, " +
                    $"SupplierId: {supplierId}, " +
                    $"SupplierCommissionPercentage: {supplierCommissionPercentage}");
```

This delists the session using the code in the `catch` block.

When the second insert fires, the `context.Entity` of the payment does not have it's BookingID property set even though it is definitely passed to the InsertPaymentAsync method:

```
        public static async Task<Payment> InsertPaymentAsync(AutoGuruDbContext db, string paymentMethodId,
            decimal amountIncGst, int? bookingId, DateTimeOffset createdAt)
        {
            var payment = new Payment
            {
                PaymentMethodId = paymentMethodId,
                AmountIncGst = amountIncGst,
                TransactionId = Guid.NewGuid().ToString(),
                TransactionStatus = "authorise_pending",
                TransactionStatusDate = createdAt,
                CreatedAt = createdAt,
                BookingID = bookingId
            };

            await db.InsertAsync(payment);

            return payment;
        }
```

If I comment out the `DelistTriggerSession` the BookingID is populated.

**Comment by anthony-keller** [comment]
image

Becomes `null` in `context.Entity`

image

**Comment by anthony-keller** [comment]
Changed the test to use a different payment method and amount

```
                // Act / Assert
                var ex1 = await Should.ThrowAsync<NotImplementedException>(() =>
                    PaymentGenerator.InsertPaymentAsync(db, PaymentMethodEnum.Afterpay.ToString(), 100m, null, effectiveAt));
                ex1.Message.ShouldBe("Couldn't find any applicable payment method for given args. " +
                    $"EffectiveAt: {effectiveAt}, PaymentMethodId: {PaymentMethodEnum.Afterpay}, BookingTypeId: , SupplierId: , SupplierCommissionPercentage: ");

                // Act / Assert
                var ex2 = await Should.ThrowAsync<NotImplementedException>(() =>
                    PaymentGenerator.InsertPaymentAsync(db, PaymentMethodEnum.Openpay.ToString(), 120m, booking.BookingID, effectiveAt));
                ex2.Message.ShouldBe("Couldn't find any applicable payment method for given args. " +
                    $"EffectiveAt: {effectiveAt}, PaymentMethodId: {PaymentMethodEnum.Afterpay}, BookingTypeId: {booking.BookingTypeID}, SupplierId: {booking.SupplierID}, SupplierCommissionPercentage: 0.00000");
```

The second `Insert` still gets the old payment values in `context.Entity` in the trigger. Without the Delist added in this PR, the values are correct.

**Comment by koenbeuk** [comment]
I've added some additional integration tests to see if I could reproduce (which I was not), here: https://github.com/koenbeuk/EntityFrameworkCore.Triggered/blob/issue-168/test/EntityFrameworkCore.Triggered.IntegrationTests/TriggerExceptions/ExceptionTestScenario.cs

In your  shared code, you do reuse the db instance between the 2 calls. Could the exception be due to the db instance still holding on on earlier data (generated during the first call)?

Keep in mind that throwing an exception in a BeforeSave trigger will not reset the state of the DbContext instance, e.g.

```csharp
var entity = new Entity();
db.Add(entity);
try {
  db.SaveChanges(); // Assume that this throws in a BeforeSaveTrigger
}
catch (Exception ex) { ... }

Console.WriteLine(db.Entry(entity).State); // Writes: Added
```

**Comment by anthony-keller** [comment]
Good point. Let me refactor it a bit and retest

**Comment by anthony-keller** [comment]
You're right - the object had the same hash code in the second call. I will run the tests today (and test https://github.com/koenbeuk/EntityFrameworkCore.Triggered/issues/156) and let you know.

Appreciate your work 💯

**Comment by koenbeuk** [comment]
Keep me posted, I've hidden the 3.2.2 release on nuget for now. Just in case there is a regression here

**Comment by anthony-keller** [comment]
@koenbeuk I have tried to reproduce the issue using the new version and have not been able to. I've not noticed any other side effects so I believe the fix is good.

**Pull request merged**